File: 백준/Gold/13392. 방법을 출력하지 않는 숫자 맞추기/방법을 출력하지 않는 숫자 맞추기.py
# 방법을 출력하지 않는 숫자 맞추기
import sys
input = sys.stdin.readline

n = int(input())
# dp[i][j] -> i번째 숫자까지 확인했을 때, 여태까지 왼쪽으로 이동횟수가 j번 일때
# target 숫자로 맞추기 위한 이동 횟수
# ex) dp[3][4] -> 3번 숫자까지 target 숫자로 맞췄을 때, 왼쪽으로 이동 횟수가 4번일때 총 이동횟수 
dp = [[1000000] * (10) for _ in range(n+1)] 
start = [0] + [int(i) for i in input().rstrip()]
target = [0] + [int(i) for i in input().rstrip()]
result = 1000000

for i in range(10):
    dp[0][i] = i

for i in range(1, n+1):
    for j in range(10):
        left = (target[i] - start[i] - j + 20) % 10
        right = 10 - left

        dp[i][j] = min(dp[i][j], dp[i-1][j] + right)
        dp[i][(j + left) % 10] = min(dp[i][(j + left) % 10], dp[i-1][j] + left)
    
for i in range(10):
    result = min(result, dp[n][i])

print(result)

File: 백준/Gold/13392. 방법을 출력하지 않는 숫자 맞추기/방법을 출력하지 않는 숫자 맞추기.java
import java.io.*;
import java.util.Arrays;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int result = 100001;
        int[][] dp = new int[n+1][10];
        int[] start = new int[n+1];
        int[] target = new int[n+1];
        String strStart = br.readLine();
        String strTarget = br.readLine();

        for (int i = 1; i <= n; i++) {
            start[i] = strStart.charAt(i - 1) - '0';
            target[i] = strTarget.charAt(i - 1) - '0';
        }

        for (int i = 1; i <= n; i++) {
            Arrays.fill(dp[i], 100001);
        }

        for (int i = 0; i < 10; i++) {
            dp[0][i] = i;
        }

        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < 10; j++) {
                int left = (target[i] - start[i] - j + 20) % 10;
                int right = 10 - left;

                dp[i][j] = Math.min(dp[i][j], dp[i-1][j] + right);
                dp[i][(j+left) % 10] = Math.min(dp[i][(j+left) % 10], dp[i-1][j] + left);
            }
        }

        for (int i = 0; i < 10; i++) {
            result = Math.min(result, dp[n][i]);
        }

        System.out.println(result);
    }
}
